### ISSUE TITLE
latest rust-analyzer consumes all memory and cpu on debian stretch and vsc 1.40.1

### ISSUE THREAD

**Issue by Dowwie** [rust-analyzer/rust-analyzer#2346]
Unfortunately, I've had to discontinue use of the rust analyzer, again, because it continues to have major system issues in that it completely consumes 16GB of memory and 100% of cpu.  I'm using Debian stretch and latest version of Visual Studio Code, v1.40.1.  This experience is based on the rust-analyzer in this repo as of this reported issue (Nov 21, 2019).  I had several weeks of no trouble at all-- maybe the last 1 to 2 months were OK-- but now, even a reinstall of the VSC extension doesn't seem to help.

**Comment by marcogroppo** [comment]
Does it happen for every project? If not, this may be a duplicate of https://github.com/rust-analyzer/rust-analyzer/issues/2284 (which in turn is due to a bug in chalk)

**Comment by Dowwie** [comment]
I love your day of the tentacle github avatar!

This experience has been within a VSC workspace that has a single cargo workspace that spans many projects.

**Comment by marcogroppo** [comment]
@Dowwie thanks! :) Unfortunately the chalk-related bug can be triggered even without control-clicking. Even opening a file with the "right" definitions can be enough (for example this one, taken from matklad's reproduction of the bug:  https://gist.github.com/marcogroppo/b52701bee35d44fca4736e2572af91f3). So I'm pretty positive that this is the same issue, but if you want to be really sure the only way is to capture a stack trace (rust-analyzer needs to be compiled with `debug=1` in the `profile.release` section in `Cargo.toml`, then you can use `gdb`).

**Comment by Dowwie** [comment]
@marcogroppo seems like a reasonable assumption that this is related to the chalk issue.  Unfortunately, I don't know how to navigate with gdb, so I'll have to wait for the chalk fix and then re-try.

**Comment by bjorn3** [comment]
@Dowwie To get a backtrace using gdb you can run:

```
$ gdb -p <pid of ra_lsp_server>
(gdb) bt all
```

You can then run use `quit` to exit gdb.

**Comment by Matthias247** [comment]
@Dowwie You can also go back to an older version temporarily if you want to continue rust-analyzer until an issue is resolved. E.g. for me ef02296b9fe2c2e42b9390a3a14ee860f08fd02c works, while some revisions after it had issues

**Comment by marcogroppo** [comment]
BTW the chalk issue has probably been resolved, but the pull request hasn't been reviewed (and merged) yet. In theory one could also replace the `chalk` dependency with the PR branch: https://github.com/rust-lang/chalk/pull/294

**Comment by Dowwie** [comment]
I'll keep an eye out for for a merge.  Thanks for the update

**Comment by Dowwie** [comment]
thanks @Matthias247  !

**Comment by marcogroppo** [comment]
> I'll keep an eye out for for a merge. Thanks for the update

`rust-analyzer` has now been updated to use the `chalk` PR branch. In theory this should have fixed this issue.

**Comment by Dowwie** [comment]
Ok!  I've reinstalled and am hoping for the best.  Thanks for the update.

**Comment by ryankurte** [comment]
this appears to still occur (unless it is a different issue with the same symptoms) using today's master (3376c08052a563a5d2db487c458972378edebf44) with vscode 0.14.1

![image](https://user-images.githubusercontent.com/860620/70028809-7c5a5400-160a-11ea-9a6d-dd678a3691ef.png)

**Comment by matklad** [comment]
@ryankurte interesting, I indeed think that we've plugged all known memory sinks recently. Could you get a stacktrace from rust-analyzer? Or, even better, are you able to create a reproducible example that shows such behavior? 

Also, am I reading this correctly that `rust-analyzer` has been running for 5 hours? Does this mean that the issue here is a slow leaking of memory, rather than a huge spike in the memory usage due to some nasty allocation heavy infinite loop?

**Comment by ryankurte** [comment]
> Could you get a stacktrace from rust-analyzer?

i had a go at `bt all` but gdb reports `No symbol "all" in current context.`. `thread apply all bt` returns [this](https://gist.github.com/ryankurte/0890371c8b0c52176fa5ec44a3cc08b8) (sorry, it's a lot...)

> are you able to create a reproducible example that shows such behavior

i'm working in a workspace across a bunch of projects, will see if i can isolate / reproduce anything useful.

> Also, am I reading this correctly that rust-analyzer has been running for 5 hours

CPU time rather than real time, the memory use builds up over 5-10 minutes then the machine starts to lock up. A sample from 20:45 (15 minutes runtime)
```
> ps aux
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
...
ryan     20654 32.1 41.4 14258808 13209576 ?   tl   20:30   4:44 ra_lsp_server
```

So CPU time is super over-reported, iirc this can happen if the task is getting suspended / resumed by the kernel at a rate that is faster than the polling interval for these stats.

**Comment by LucioFranco** [comment]
I am happy to look into this a bit more later. But I wanted to chime in that I also had this bug across all three OSes and was bringing my computer to a halt. For me I can reproduce it on a 16gb machine by running vscode + ra with `timberio/vector` then attempting to run `cargo test` in a shell. I hit the limits of my system during linking even if ra has only just loaded the workspace.

**Comment by jeromegn** [comment]
Getting the same issues on my project. It uses about 360 crates. I'm running Arch with linux kernel 5.3, on a 6-core, 16GB memory laptop.

Love rust-analyzer, once this issue is resolved it's going to be nearly perfect!

**Comment by ryankurte** [comment]
hmm, i have also been running `cargo check` in a terminal alongside it (but have not enabled r-a's check functionality)

**Comment by Dowwie** [comment]
I don't want to jinx this but ever since @marcogroppo announced the update to master, I re-installed rust-analyzer in Debian stretch and latest VSC and have had no issues.  I didn't want to report this too soon but it's been 9 days of flawless activity.

**Comment by Matthias247** [comment]
For me this got definitely worse with recent updates. rust-analyzer regularly hits 100% CPU utilization after a short time of use (5-30min): 

<img width="434" alt="Screen Shot 2019-12-09 at 12 25 33" src="https://user-images.githubusercontent.com/6330334/70470067-1c870180-1a7f-11ea-9cdc-383cb542ba7c.png">
<img width="421" alt="Screen Shot 2019-12-09 at 12 25 44" src="https://user-images.githubusercontent.com/6330334/70470094-26106980-1a7f-11ea-9e0f-ecb0b6d04c5d.png">

I am on commit 9e551d5452232d1d44e4bba754392beaa3a7628f right now

**Comment by lnicola** [comment]
If it's still running, can you try running `perf top -p $(pidof ra_lsp_server)`? The debug info is disabled by default, but it still seems to report something for me.

**Comment by Matthias247** [comment]
I'm on MacOS, so there is no `perf`. I tried "Instruments", but it doesn't show anything. 
I just reported into this issue, since this is the only high resource utilization issue that is currently open.

**Comment by lnicola** [comment]
You might want to try after #2516 gets merged. I had a project where `ra_cli analysis-stats` took 153 seconds and now it's down to 30.

**Comment by Matthias247** [comment]
Thanks @lnicola 

I tried updating, but it still has the same issue.
I however was now stable to capture something using instruments.

In heaviest stack trace, it shows:

```
  66  22539.0  ra_lsp_server (82749) :0
  65  18100.0  std::sys::unix::thread::Thread::new::thread_start::h785fcf8d81bd0f52  0x4a6e2 :0
  64 libsystem_pthread.dylib 18100.0  thread_start
  63 libsystem_pthread.dylib 18100.0  _pthread_start
  62 libsystem_pthread.dylib 18100.0  _pthread_body
  61 ra_lsp_server 18100.0  std::sys::unix::thread::Thread::new::thread_start::h785fcf8d81bd0f52
  60 ra_lsp_server 18100.0  _$LT$alloc..boxed..Box$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$A$GT$$GT$::call_once::hd1afe4d791e3679f
  59 ra_lsp_server 18100.0  core::ops::function::FnOnce::call_once$u7b$$u7b$vtable.shim$u7d$$u7d$::hdec869ac48a2a287
  58 ra_lsp_server 18100.0  __rust_maybe_catch_panic
  57 ra_lsp_server 18100.0  std::sys_common::backtrace::__rust_begin_short_backtrace::h2c0eeee5042731aa (.llvm.6494389862530504639)
  56 ra_lsp_server 18100.0  _$LT$F$u20$as$u20$threadpool..FnBox$GT$::call_box::h4f26c3f2bd35b2d6
  55 ra_lsp_server 17772.0  ra_lsp_server::main_loop::handlers::publish_diagnostics::hf82faed240bbda9b
  54 ra_lsp_server 17772.0  ra_ide::Analysis::diagnostics::hfae9ee480c170ef7
  53 ra_lsp_server 17772.0  ra_db::CheckCanceled::catch_canceled::ha09639be16fa9d94
  52 ra_lsp_server 17772.0  __rust_maybe_catch_panic
  51 ra_lsp_server 17772.0  std::panicking::try::do_call::h774adf3d7b9a1679 (.llvm.18197795992867467660)
  50 ra_lsp_server 17772.0  ra_ide::diagnostics::diagnostics::h0226cd3ea61422a0
  49 ra_lsp_server 17767.0  ra_hir::code_model::Module::diagnostics::h0785fc70ac42f123
  48 ra_lsp_server 17767.0  ra_hir::code_model::Function::diagnostics::hb7d60bc2a32d0473
  47 ra_lsp_server 17767.0  _$LT$salsa..derived..DerivedStorage$LT$DB$C$Q$C$MP$GT$$u20$as$u20$salsa..plumbing..QueryStorageOps$LT$DB$C$Q$GT$$GT$::try_fetch::h3ab9e62720e5516e
  46 ra_lsp_server 17767.0  salsa::derived::slot::Slot$LT$DB$C$Q$C$MP$GT$::read::h2c0afa16fcbc0767
  45 ra_lsp_server 17765.0  salsa::derived::slot::Slot$LT$DB$C$Q$C$MP$GT$::read_upgrade::h7e503cebf140c61f
  44 ra_lsp_server 17765.0  _$LT$salsa..derived..slot..Slot$LT$DB$C$Q$C$MP$GT$$u20$as$u20$salsa..dependency..DatabaseSlot$LT$DB$GT$$GT$::maybe_changed_since::hfd1110eafe8bc15e
  43 ra_lsp_server 17764.0  salsa::derived::slot::Slot$LT$DB$C$Q$C$MP$GT$::read_upgrade::hd28c794787523283
  42 ra_lsp_server 17764.0  salsa::runtime::Runtime$LT$DB$GT$::execute_query_implementation::h3da5ee46fc185714
  41 ra_lsp_server 17764.0  ra_hir_ty::traits::trait_solve_query::h7cf471a00c07bbd5
  40 ra_lsp_server 17764.0  chalk_engine::forest::Forest$LT$C$GT$::solve::heecc7bc5884f0ef1
  39 ra_lsp_server 17710.0  chalk_solve::solve::slg::aggregate::_$LT$impl$u20$chalk_engine..context..AggregateOps$LT$chalk_solve..solve..slg..SlgContext$LT$TF$GT$$GT$$u20$for$u20$chalk_solve..solve..slg..SlgContextOps$LT$TF$GT$$GT$::make_solution::h834a40ca1ba98b92
  38 ra_lsp_server 17710.0  _$LT$chalk_engine..forest..ForestSolver$LT$C$C$CO$GT$$u20$as$u20$chalk_engine..context..AnswerStream$LT$C$GT$$GT$::peek_answer::hc1dbd2bac96a626c
  37 ra_lsp_server 17704.0  chalk_engine::logic::_$LT$impl$u20$chalk_engine..forest..Forest$LT$C$GT$$GT$::ensure_root_answer::h8637fcbd36ed6a37
  36 ra_lsp_server 16466.0  chalk_engine::logic::_$LT$impl$u20$chalk_engine..forest..Forest$LT$C$GT$$GT$::get_or_create_table_for_ucanonical_goal::ha2452aa1765187bb (.llvm.15219803596393920156)
  35 ra_lsp_server 16248.0  chalk_engine::logic::_$LT$impl$u20$chalk_engine..forest..Forest$LT$C$GT$$GT$::push_initial_strands_instantiated::hbcb2aa5857d1b1c2
  34 ra_lsp_server 13205.0  _$LT$chalk_solve..solve..slg..SlgContextOps$LT$TF$GT$$u20$as$u20$chalk_engine..context..ContextOps$LT$chalk_solve..solve..slg..SlgContext$LT$TF$GT$$GT$$GT$::program_clauses::h7d3946d01e66280e
  33 ra_lsp_server 13197.0  chalk_solve::clauses::program_clauses_for_goal::h100574e82ab49014
  32 ra_lsp_server 7597.0  _$LT$chalk_rust_ir..AssociatedTyValue$LT$TF$GT$$u20$as$u20$chalk_solve..clauses..program_clauses..ToProgramClauses$LT$TF$GT$$GT$::to_program_clauses::h1eb26983f36338ac
  31 ra_lsp_server 6964.0  chalk_solve::clauses::builder::ClauseBuilder$LT$TF$GT$::push_binders::h972d2cd1888a132a
  30 ra_lsp_server 2524.0  chalk_solve::split::Split::impl_parameters_and_projection_from_associated_ty_value::hcd0a1bea38ea81d4
  29 ra_lsp_server 1039.0  chalk_ir::Binders$LT$T$GT$::substitute::hbd125a8c09b7ed3d
  28 ra_lsp_server 1017.0  core::iter::adapters::process_results::hc9124a622e23470e
  27 ra_lsp_server 1001.0  _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$alloc..vec..SpecExtend$LT$T$C$I$GT$$GT$::from_iter::h19005578466607bc
  26 ra_lsp_server 757.0  _$LT$core..iter..adapters..ResultShunt$LT$I$C$E$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$::next::hb0c5e0f43d34d676
  25 ra_lsp_server 713.0  _$LT$chalk_ir..Parameter$LT$TF$GT$$u20$as$u20$chalk_ir..fold..Fold$LT$TF$C$TTF$GT$$GT$::fold_with::h89fe22304fa9cdf6
  24 ra_lsp_server 567.0  _$LT$T$u20$as$u20$chalk_ir..fold..TypeFolder$LT$TF$C$TTF$GT$$GT$::fold_ty::h4c3bff360afc489c
  23 ra_lsp_server 565.0  chalk_ir::fold::super_fold_ty::h8555d4e029e0b486
  22 ra_lsp_server 551.0  _$LT$chalk_ir..ApplicationTy$LT$TF$GT$$u20$as$u20$chalk_ir..fold..Fold$LT$TF$C$TTF$GT$$GT$::fold_with::h50422d15b3b6c940
  21 ra_lsp_server 494.0  core::iter::adapters::process_results::hc9124a622e23470e
  20 ra_lsp_server 480.0  _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$alloc..vec..SpecExtend$LT$T$C$I$GT$$GT$::from_iter::h19005578466607bc
  19 ra_lsp_server 228.0  _$LT$core..iter..adapters..ResultShunt$LT$I$C$E$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$::next::hb0c5e0f43d34d676
  18 ra_lsp_server 198.0  _$LT$chalk_ir..Parameter$LT$TF$GT$$u20$as$u20$chalk_ir..fold..Fold$LT$TF$C$TTF$GT$$GT$::fold_with::h89fe22304fa9cdf6
  17 ra_lsp_server 129.0  _$LT$T$u20$as$u20$chalk_ir..fold..TypeFolder$LT$TF$C$TTF$GT$$GT$::fold_ty::h4c3bff360afc489c
  16 ra_lsp_server 124.0  chalk_ir::fold::super_fold_ty::h8555d4e029e0b486
  15 ra_lsp_server 69.0  _$LT$chalk_ir..ApplicationTy$LT$TF$GT$$u20$as$u20$chalk_ir..fold..Fold$LT$TF$C$TTF$GT$$GT$::fold_with::h50422d15b3b6c940
  14 ra_lsp_server 64.0  core::iter::adapters::process_results::hc9124a622e23470e
  13 ra_lsp_server 62.0  _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$alloc..vec..SpecExtend$LT$T$C$I$GT$$GT$::from_iter::h19005578466607bc
  12 ra_lsp_server 36.0  _$LT$core..iter..adapters..ResultShunt$LT$I$C$E$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$::next::hb0c5e0f43d34d676
  11 ra_lsp_server 32.0  _$LT$chalk_ir..Parameter$LT$TF$GT$$u20$as$u20$chalk_ir..fold..Fold$LT$TF$C$TTF$GT$$GT$::fold_with::h89fe22304fa9cdf6
  10 ra_lsp_server 26.0  _$LT$T$u20$as$u20$chalk_ir..fold..TypeFolder$LT$TF$C$TTF$GT$$GT$::fold_ty::h4c3bff360afc489c
   9 ra_lsp_server 26.0  chalk_ir::fold::super_fold_ty::h8555d4e029e0b486
   8 ra_lsp_server 21.0  _$LT$chalk_ir..ApplicationTy$LT$TF$GT$$u20$as$u20$chalk_ir..fold..Fold$LT$TF$C$TTF$GT$$GT$::fold_with::h50422d15b3b6c940
   7 ra_lsp_server 21.0  core::iter::adapters::process_results::hc9124a622e23470e
   6 ra_lsp_server 21.0  _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$alloc..vec..SpecExtend$LT$T$C$I$GT$$GT$::from_iter::h19005578466607bc
   5 libsystem_malloc.dylib 10.0  realloc
   4 libsystem_malloc.dylib 10.0  malloc_zone_realloc
   3 libsystem_malloc.dylib 9.0  szone_realloc
   2 libsystem_malloc.dylib 2.0  szone_malloc_should_clear
   1 libsystem_malloc.dylib 2.0  tiny_malloc_should_clear
   0 libsystem_malloc.dylib 1.0  tiny_malloc_from_free_list
````

**Comment by Matthias247** [comment]
<details><summary>Here is another view of the trace</summary>

```
22.54 s  100.0%	0 s	 	ra_lsp_server (82749)
18.10 s   80.3%	0 s	 	 std::sys::unix::thread::Thread::new::thread_start::h785fcf8d81bd0f52  0x4a6e2
18.10 s   80.3%	0 s	 	  thread_start
18.10 s   80.3%	0 s	 	   _pthread_start
18.10 s   80.3%	0 s	 	    _pthread_body
18.10 s   80.3%	0 s	 	     std::sys::unix::thread::Thread::new::thread_start::h785fcf8d81bd0f52
18.10 s   80.3%	0 s	 	      _$LT$alloc..boxed..Box$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$A$GT$$GT$::call_once::hd1afe4d791e3679f
18.10 s   80.3%	0 s	 	       core::ops::function::FnOnce::call_once$u7b$$u7b$vtable.shim$u7d$$u7d$::hdec869ac48a2a287
18.10 s   80.3%	0 s	 	        __rust_maybe_catch_panic
18.10 s   80.3%	0 s	 	         std::sys_common::backtrace::__rust_begin_short_backtrace::h2c0eeee5042731aa (.llvm.6494389862530504639)
18.10 s   80.3%	0 s	 	          _$LT$F$u20$as$u20$threadpool..FnBox$GT$::call_box::h4f26c3f2bd35b2d6
17.77 s   78.8%	0 s	 	           ra_lsp_server::main_loop::handlers::publish_diagnostics::hf82faed240bbda9b
17.77 s   78.8%	0 s	 	            ra_ide::Analysis::diagnostics::hfae9ee480c170ef7
17.77 s   78.8%	0 s	 	             ra_db::CheckCanceled::catch_canceled::ha09639be16fa9d94
17.77 s   78.8%	0 s	 	              __rust_maybe_catch_panic
17.77 s   78.8%	0 s	 	               std::panicking::try::do_call::h774adf3d7b9a1679 (.llvm.18197795992867467660)
17.77 s   78.8%	0 s	 	                ra_ide::diagnostics::diagnostics::h0226cd3ea61422a0
17.77 s   78.8%	0 s	 	                 ra_hir::code_model::Module::diagnostics::h0785fc70ac42f123
17.77 s   78.8%	0 s	 	                  ra_hir::code_model::Function::diagnostics::hb7d60bc2a32d0473
17.77 s   78.8%	0 s	 	                   _$LT$salsa..derived..DerivedStorage$LT$DB$C$Q$C$MP$GT$$u20$as$u20$salsa..plumbing..QueryStorageOps$LT$DB$C$Q$GT$$GT$::try_fetch::h3ab9e62720e5516e
17.77 s   78.8%	0 s	 	                    salsa::derived::slot::Slot$LT$DB$C$Q$C$MP$GT$::read::h2c0afa16fcbc0767
17.77 s   78.8%	0 s	 	                     salsa::derived::slot::Slot$LT$DB$C$Q$C$MP$GT$::read_upgrade::h7e503cebf140c61f
17.77 s   78.8%	1.00 ms	 	                      _$LT$salsa..derived..slot..Slot$LT$DB$C$Q$C$MP$GT$$u20$as$u20$salsa..dependency..DatabaseSlot$LT$DB$GT$$GT$::maybe_changed_since::hfd1110eafe8bc15e
17.76 s   78.8%	0 s	 	                       salsa::derived::slot::Slot$LT$DB$C$Q$C$MP$GT$::read_upgrade::hd28c794787523283
17.76 s   78.8%	0 s	 	                        salsa::runtime::Runtime$LT$DB$GT$::execute_query_implementation::h3da5ee46fc185714
17.76 s   78.8%	0 s	 	                         ra_hir_ty::traits::trait_solve_query::h7cf471a00c07bbd5
17.76 s   78.8%	0 s	 	                          chalk_engine::forest::Forest$LT$C$GT$::solve::heecc7bc5884f0ef1
17.71 s   78.5%	0 s	 	                           chalk_solve::solve::slg::aggregate::_$LT$impl$u20$chalk_engine..context..AggregateOps$LT$chalk_solve..solve..slg..SlgContext$LT$TF$GT$$GT$$u20$for$u20$chalk_solve..solve..slg..SlgContextOps$LT$TF$GT$$GT$::make_solution::h834a40ca1ba98b92
17.71 s   78.5%	0 s	 	                            _$LT$chalk_engine..forest..ForestSolver$LT$C$C$CO$GT$$u20$as$u20$chalk_engine..context..AnswerStream$LT$C$GT$$GT$::peek_answer::hc1dbd2bac96a626c
17.70 s   78.5%	32.00 ms	 	                             chalk_engine::logic::_$LT$impl$u20$chalk_engine..forest..Forest$LT$C$GT$$GT$::ensure_root_answer::h8637fcbd36ed6a37
16.47 s   73.0%	2.00 ms	 	                              chalk_engine::logic::_$LT$impl$u20$chalk_engine..forest..Forest$LT$C$GT$$GT$::get_or_create_table_for_ucanonical_goal::ha2452aa1765187bb (.llvm.15219803596393920156)
16.25 s   72.0%	20.00 ms	 	                               chalk_engine::logic::_$LT$impl$u20$chalk_engine..forest..Forest$LT$C$GT$$GT$::push_initial_strands_instantiated::hbcb2aa5857d1b1c2
13.21 s   58.5%	3.00 ms	 	                                _$LT$chalk_solve..solve..slg..SlgContextOps$LT$TF$GT$$u20$as$u20$chalk_engine..context..ContextOps$LT$chalk_solve..solve..slg..SlgContext$LT$TF$GT$$GT$$GT$::program_clauses::h7d3946d01e66280e
13.20 s   58.5%	215.00 ms	 	                                 chalk_solve::clauses::program_clauses_for_goal::h100574e82ab49014
7.60 s   33.7%	111.00 ms	 	                                  _$LT$chalk_rust_ir..AssociatedTyValue$LT$TF$GT$$u20$as$u20$chalk_solve..clauses..program_clauses..ToProgramClauses$LT$TF$GT$$GT$::to_program_clauses::h1eb26983f36338ac
6.96 s   30.8%	232.00 ms	 	                                   chalk_solve::clauses::builder::ClauseBuilder$LT$TF$GT$::push_binders::h972d2cd1888a132a
2.52 s   11.1%	91.00 ms	 	                                    chalk_solve::split::Split::impl_parameters_and_projection_from_associated_ty_value::hcd0a1bea38ea81d4
1.04 s    4.6%	22.00 ms	 	                                     chalk_ir::Binders$LT$T$GT$::substitute::hbd125a8c09b7ed3d
1.02 s    4.5%	11.00 ms	 	                                      core::iter::adapters::process_results::hc9124a622e23470e
1.00 s    4.4%	41.00 ms	 	                                       _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$alloc..vec..SpecExtend$LT$T$C$I$GT$$GT$::from_iter::h19005578466607bc
757.00 ms    3.3%	43.00 ms	 	                                        _$LT$core..iter..adapters..ResultShunt$LT$I$C$E$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$::next::hb0c5e0f43d34d676
713.00 ms    3.1%	145.00 ms	 	                                         _$LT$chalk_ir..Parameter$LT$TF$GT$$u20$as$u20$chalk_ir..fold..Fold$LT$TF$C$TTF$GT$$GT$::fold_with::h89fe22304fa9cdf6
567.00 ms    2.5%	2.00 ms	 	                                          _$LT$T$u20$as$u20$chalk_ir..fold..TypeFolder$LT$TF$C$TTF$GT$$GT$::fold_ty::h4c3bff360afc489c
565.00 ms    2.5%	14.00 ms	 	                                           chalk_ir::fold::super_fold_ty::h8555d4e029e0b486
551.00 ms    2.4%	40.00 ms	 	                                            _$LT$chalk_ir..ApplicationTy$LT$TF$GT$$u20$as$u20$chalk_ir..fold..Fold$LT$TF$C$TTF$GT$$GT$::fold_with::h50422d15b3b6c940
494.00 ms    2.1%	11.00 ms	 	                                             core::iter::adapters::process_results::hc9124a622e23470e
480.00 ms    2.1%	32.00 ms	 	                                              _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$alloc..vec..SpecExtend$LT$T$C$I$GT$$GT$::from_iter::h19005578466607bc
228.00 ms    1.0%	30.00 ms	 	                                               _$LT$core..iter..adapters..ResultShunt$LT$I$C$E$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$::next::hb0c5e0f43d34d676
198.00 ms    0.8%	69.00 ms	 	                                                _$LT$chalk_ir..Parameter$LT$TF$GT$$u20$as$u20$chalk_ir..fold..Fold$LT$TF$C$TTF$GT$$GT$::fold_with::h89fe22304fa9cdf6
129.00 ms    0.5%	4.00 ms	 	                                                 _$LT$T$u20$as$u20$chalk_ir..fold..TypeFolder$LT$TF$C$TTF$GT$$GT$::fold_ty::h4c3bff360afc489c
124.00 ms    0.5%	13.00 ms	 	                                                  chalk_ir::fold::super_fold_ty::h8555d4e029e0b486
69.00 ms    0.3%	4.00 ms	 	                                                   _$LT$chalk_ir..ApplicationTy$LT$TF$GT$$u20$as$u20$chalk_ir..fold..Fold$LT$TF$C$TTF$GT$$GT$::fold_with::h50422d15b3b6c940
64.00 ms    0.2%	2.00 ms	 	                                                    core::iter::adapters::process_results::hc9124a622e23470e
62.00 ms    0.2%	2.00 ms	 	                                                     _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$alloc..vec..SpecExtend$LT$T$C$I$GT$$GT$::from_iter::h19005578466607bc
36.00 ms    0.1%	4.00 ms	 	                                                      _$LT$core..iter..adapters..ResultShunt$LT$I$C$E$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$::next::hb0c5e0f43d34d676
32.00 ms    0.1%	6.00 ms	 	                                                       _$LT$chalk_ir..Parameter$LT$TF$GT$$u20$as$u20$chalk_ir..fold..Fold$LT$TF$C$TTF$GT$$GT$::fold_with::h89fe22304fa9cdf6
26.00 ms    0.1%	0 s	 	                                                        _$LT$T$u20$as$u20$chalk_ir..fold..TypeFolder$LT$TF$C$TTF$GT$$GT$::fold_ty::h4c3bff360afc489c
26.00 ms    0.1%	1.00 ms	 	                                                         chalk_ir::fold::super_fold_ty::h8555d4e029e0b486
21.00 ms    0.0%	0 s	 	                                                          _$LT$chalk_ir..ApplicationTy$LT$TF$GT$$u20$as$u20$chalk_ir..fold..Fold$LT$TF$C$TTF$GT$$GT$::fold_with::h50422d15b3b6c940
21.00 ms    0.0%	0 s	 	                                                           core::iter::adapters::process_results::hc9124a622e23470e
21.00 ms    0.0%	0 s	 	                                                            _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$alloc..vec..SpecExtend$LT$T$C$I$GT$$GT$::from_iter::h19005578466607bc
10.00 ms    0.0%	0 s	 	                                                             realloc
10.00 ms    0.0%	0 s	 	                                                              malloc_zone_realloc
9.00 ms    0.0%	1.00 ms	 	                                                               szone_realloc
2.00 ms    0.0%	2.00 ms	 	                                                                _platform_memmove$VARIANT$Haswell
2.00 ms    0.0%	0 s	 	                                                                szone_malloc_should_clear
2.00 ms    0.0%	1.00 ms	 	                                                                 tiny_malloc_should_clear
2.00 ms    0.0%	2.00 ms	 	                                                                tiny_size
1.00 ms    0.0%	1.00 ms	 	                                                                tiny_try_realloc_in_place
1.00 ms    0.0%	0 s	 	                                                                free_tiny
1.00 ms    0.0%	1.00 ms	 	                                                               szone_free
7.00 ms    0.0%	1.00 ms	 	                                                             _$LT$core..iter..adapters..ResultShunt$LT$I$C$E$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$::next::hb0c5e0f43d34d676
4.00 ms    0.0%	0 s	 	                                                             malloc
4.00 ms    0.0%	0 s	 	                                                              malloc_zone_malloc
3.00 ms    0.0%	0 s	 	                                                               szone_malloc_should_clear
3.00 ms    0.0%	0 s	 	                                                                tiny_malloc_should_clear
2.00 ms    0.0%	2.00 ms	 	                                                                 tiny_malloc_from_free_list
1.00 ms    0.0%	1.00 ms	 	                                                                 set_tiny_meta_header_in_use
1.00 ms    0.0%	1.00 ms	 	                                                               default_zone_malloc
4.00 ms    0.0%	3.00 ms	 	                                                          _$LT$chalk_ir..fold..subst..Subst$LT$TF$GT$$u20$as$u20$chalk_ir..fold..FreeVarFolder$LT$TF$GT$$GT$::fold_free_var_ty::h7cf0eaf896a042e4
15.00 ms    0.0%	0 s	 	                                                      realloc
13.00 ms    0.0%	0 s	 	                                                       malloc_zone_realloc
13.00 ms    0.0%	1.00 ms	 	                                                        szone_realloc
4.00 ms    0.0%	1.00 ms	 	                                                         szone_malloc_should_clear
3.00 ms    0.0%	1.00 ms	 	                                                          tiny_malloc_should_clear
2.00 ms    0.0%	2.00 ms	 	                                                           tiny_malloc_from_free_list
4.00 ms    0.0%	4.00 ms	 	                                                         tiny_try_realloc_in_place
2.00 ms    0.0%	2.00 ms	 	                                                         tiny_size
1.00 ms    0.0%	1.00 ms	 	                                                         free_tiny
1.00 ms    0.0%	1.00 ms	 	                                                         set_tiny_meta_header_in_use
2.00 ms    0.0%	2.00 ms	 	                                                       tiny_size
9.00 ms    0.0%	0 s	 	                                                      malloc
1.00 ms    0.0%	1.00 ms	 	                                                    _$LT$chalk_ir..family..ChalkIr$u20$as$u20$chalk_ir..family..TypeFamily$GT$::intern_ty::h35af7c51490f2ffa
42.00 ms    0.1%	11.00 ms	 	                                                   _$LT$chalk_ir..fold..subst..Subst$LT$TF$GT$$u20$as$u20$chalk_ir..fold..FreeVarFolder$LT$TF$GT$$GT$::fold_free_var_ty::h7cf0eaf896a042e4
1.00 ms    0.0%	0 s	 	                                                  <Unknown Address>
114.00 ms    0.5%	1.00 ms	 	                                               malloc
106.00 ms    0.4%	1.00 ms	 	                                               realloc
93.00 ms    0.4%	5.00 ms	 	                                                malloc_zone_realloc
80.00 ms    0.3%	5.00 ms	 	                                                 szone_realloc
29.00 ms    0.1%	1.00 ms	 	                                                  szone_malloc_should_clear
28.00 ms    0.1%	13.00 ms	 	                                                   tiny_malloc_should_clear
11.00 ms    0.0%	11.00 ms	 	                                                    tiny_malloc_from_free_list
4.00 ms    0.0%	4.00 ms	 	                                                    set_tiny_meta_header_in_use
16.00 ms    0.0%	14.00 ms	 	                                                  tiny_try_realloc_in_place
10.00 ms    0.0%	10.00 ms	 	                                                  free_tiny
10.00 ms    0.0%	10.00 ms	 	                                                  _platform_memmove$VARIANT$Haswell
4.00 ms    0.0%	4.00 ms	 	                                                  tiny_size
4.00 ms    0.0%	4.00 ms	 	                                                  set_tiny_meta_header_in_use
2.00 ms    0.0%	2.00 ms	 	                                                  szone_size
3.00 ms    0.0%	3.00 ms	 	                                                 default_zone_realloc
2.00 ms    0.0%	2.00 ms	 	                                                 DYLD-STUB$$_platform_memmove
2.00 ms    0.0%	2.00 ms	 	                                                 szone_free
1.00 ms    0.0%	1.00 ms	 	                                                 DYLD-STUB$$_platform_memmove
12.00 ms    0.0%	12.00 ms	 	                                                tiny_size
2.00 ms    0.0%	0 s	 	                                              <Unknown Address>
1.00 ms    0.0%	1.00 ms	 	                                              __rust_alloc
17.00 ms    0.0%	17.00 ms	 	                                             _$LT$chalk_ir..family..ChalkIr$u20$as$u20$chalk_ir..family..TypeFamily$GT$::intern_ty::h35af7c51490f2ffa
1.00 ms    0.0%	0 s	 	                                          <Unknown Address>
1.00 ms    0.0%	0 s	 	                                         <Unknown Address>
164.00 ms    0.7%	5.00 ms	 	                                        malloc
39.00 ms    0.1%	0 s	 	                                        realloc
36.00 ms    0.1%	2.00 ms	 	                                         malloc_zone_realloc
31.00 ms    0.1%	2.00 ms	 	                                          szone_realloc
13.00 ms    0.0%	13.00 ms	 	                                           tiny_try_realloc_in_place
5.00 ms    0.0%	1.00 ms	 	                                           szone_malloc_should_clear
4.00 ms    0.0%	4.00 ms	 	                                            tiny_malloc_should_clear
4.00 ms    0.0%	4.00 ms	 	                                           _platform_memmove$VARIANT$Haswell
3.00 ms    0.0%	3.00 ms	 	                                           tiny_size
2.00 ms    0.0%	2.00 ms	 	                                           free_tiny
2.00 ms    0.0%	2.00 ms	 	                                           set_tiny_meta_header_in_use
1.00 ms    0.0%	1.00 ms	 	                                          DYLD-STUB$$_platform_memmove
1.00 ms    0.0%	1.00 ms	 	                                          szone_good_size
1.00 ms    0.0%	1.00 ms	 	                                          szone_free
3.00 ms    0.0%	3.00 ms	 	                                         tiny_size
3.00 ms    0.0%	3.00 ms	 	                                       __rdl_alloc
2.00 ms    0.0%	2.00 ms	 	                                       __rust_alloc
475.00 ms    2.1%	18.00 ms	 	                                     ra_hir_ty::traits::chalk::_$LT$impl$u20$chalk_solve..RustIrDatabase$LT$chalk_ir..family..ChalkIr$GT$$u20$for$u20$ra_hir_ty..traits..ChalkContext$LT$DB$GT$$GT$::impl_datum::h0d797a9d82263194
442.00 ms    1.9%	24.00 ms	 	                                     _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$alloc..vec..SpecExtend$LT$T$C$I$GT$$GT$::from_iter::ha0a1fb8b6d1ad082
157.00 ms    0.6%	98.00 ms	 	                                     free_tiny
87.00 ms    0.3%	33.00 ms	 	                                     free
81.00 ms    0.3%	81.00 ms	 	                                     _platform_memmove$VARIANT$Haswell
69.00 ms    0.3%	1.00 ms	 	                                     malloc
39.00 ms    0.1%	20.00 ms	 	                                     core::ptr::real_drop_in_place::haa583186e12cccc4 (.llvm.7017935223121680485)
14.00 ms    0.0%	14.00 ms	 	                                     default_zone_free_definite_size
9.00 ms    0.0%	9.00 ms	 	                                     _$LT$chalk_macros..DEBUG_ENABLED$u20$as$u20$core..ops..deref..Deref$GT$::deref::hfacdb13209b5fe1b
9.00 ms    0.0%	9.00 ms	 	                                     chalk_macros::Indent::new::h1ed049dec3cf72e8
4.00 ms    0.0%	0 s	 	                                     <Unknown Address>
2.00 ms    0.0%	2.00 ms	 	                                     alloc::raw_vec::RawVec$LT$T$C$A$GT$::reserve::h4e01024bf77f5d4b
2.00 ms    0.0%	2.00 ms	 	                                     __rdl_alloc
1.00 ms    0.0%	1.00 ms	 	                                     _$LT$chalk_macros..Indent$u20$as$u20$core..ops..drop..Drop$GT$::drop::h89a0511c17e39f14
1.00 ms    0.0%	1.00 ms	 	                                     __rust_alloc
1.00 ms    0.0%	1.00 ms	 	                                     __rdl_dealloc
1.00 ms    0.0%	1.00 ms	 	                                     core::slice::_$LT$impl$u20$$u5b$T$u5d$$GT$::copy_from_slice::hbde64a118b2fe800
1.83 s    8.1%	29.00 ms	 	                                    _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$alloc..vec..SpecExtend$LT$T$C$I$GT$$GT$::from_iter::he67ed9f4974c6a2e
776.00 ms    3.4%	15.00 ms	 	                                    chalk_ir::Binders$LT$T$GT$::substitute::h77e1c6b1b7bf8956
543.00 ms    2.4%	25.00 ms	 	                                    malloc
290.00 ms    1.2%	23.00 ms	 	                                    alloc::vec::Vec$LT$T$GT$::extend_from_slice::h9f9ec3887999902f
283.00 ms    1.2%	136.00 ms	 	                                    free_tiny
152.00 ms    0.6%	152.00 ms	 	                                    _platform_memmove$VARIANT$Haswell
99.00 ms    0.4%	42.00 ms	 	                                    free
69.00 ms    0.3%	0 s	 	                                    alloc::raw_vec::RawVec$LT$T$C$A$GT$::reserve::h474aa6550149bc73
39.00 ms    0.1%	18.00 ms	 	                                    core::ptr::real_drop_in_place::haa583186e12cccc4 (.llvm.7017935223121680485)
32.00 ms    0.1%	14.00 ms	 	                                    _$LT$core..iter..adapters..Map$LT$I$C$F$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$::fold::h47b514b8d2ee6f7a
23.00 ms    0.1%	0 s	 	                                    <Unknown Address>
17.00 ms    0.0%	15.00 ms	 	                                    alloc::raw_vec::RawVec$LT$T$C$A$GT$::reserve::h4e01024bf77f5d4b
13.00 ms    0.0%	13.00 ms	 	                                    szone_free_definite_size
12.00 ms    0.0%	5.00 ms	 	                                    alloc::raw_vec::RawVec$LT$T$C$A$GT$::reserve::h985b89ec9f925cce
12.00 ms    0.0%	12.00 ms	 	                                    default_zone_free_definite_size
5.00 ms    0.0%	5.00 ms	 	                                    __rdl_dealloc
5.00 ms    0.0%	5.00 ms	 	                                    core::slice::_$LT$impl$u20$$u5b$T$u5d$$GT$::copy_from_slice::hbde64a118b2fe800
4.00 ms    0.0%	4.00 ms	 	                                    _$LT$chalk_macros..DEBUG_ENABLED$u20$as$u20$core..ops..deref..Deref$GT$::deref::hfacdb13209b5fe1b
2.00 ms    0.0%	2.00 ms	 	                                    __rust_dealloc
2.00 ms    0.0%	2.00 ms	 	                                    __rdl_alloc
1.00 ms    0.0%	1.00 ms	 	                                    __rust_realloc
1.00 ms    0.0%	1.00 ms	 	                                    DYLD-STUB$$memcpy
1.00 ms    0.0%	1.00 ms	 	                                    DYLD-STUB$$malloc
1.00 ms    0.0%	1.00 ms	 	                                    __rust_alloc
255.00 ms    1.1%	7.00 ms	 	                                   ra_hir_ty::traits::chalk::_$LT$impl$u20$chalk_solve..RustIrDatabase$LT$chalk_ir..family..ChalkIr$GT$$u20$for$u20$ra_hir_ty..traits..ChalkContext$LT$DB$GT$$GT$::associated_ty_data::h81ff9321b764839d
233.00 ms    1.0%	7.00 ms	 	                                   ra_hir_ty::traits::chalk::_$LT$impl$u20$chalk_solve..RustIrDatabase$LT$chalk_ir..family..ChalkIr$GT$$u20$for$u20$ra_hir_ty..traits..ChalkContext$LT$DB$GT$$GT$::impl_datum::h0d797a9d82263194
20.00 ms    0.0%	0 s	 	                                   <Unknown Address>
5.00 ms    0.0%	5.00 ms	 	                                   __rdl_alloc
4.00 ms    0.0%	4.00 ms	 	                                   __rdl_dealloc
2.00 ms    0.0%	2.00 ms	 	                                   free
2.00 ms    0.0%	2.00 ms	 	                                   szone_free_definite_size
1.00 ms    0.0%	1.00 ms	 	                                   __rust_alloc
1.48 s    6.5%	42.00 ms	 	                                  alloc::vec::Vec$LT$T$GT$::retain::h7528e9b48b4e7922
863.00 ms    3.8%	7.00 ms	 	                                  chalk_solve::clauses::env_elaborator::elaborate_env_clauses::h99f0d5a9262783e3
799.00 ms    3.5%	8.00 ms	 	                                  ra_hir_ty::traits::chalk::_$LT$impl$u20$chalk_solve..RustIrDatabase$LT$chalk_ir..family..ChalkIr$GT$$u20$for$u20$ra_hir_ty..traits..ChalkContext$LT$DB$GT$$GT$::impl_datum::h0d797a9d82263194
735.00 ms    3.2%	7.00 ms	 	                                  ra_hir_ty::traits::chalk::_$LT$impl$u20$chalk_solve..RustIrDatabase$LT$chalk_ir..family..ChalkIr$GT$$u20$for$u20$ra_hir_ty..traits..ChalkContext$LT$DB$GT$$GT$::associated_ty_value::h0b10268faf2a3355
439.00 ms    1.9%	1.00 ms	 	                                  ra_hir_ty::traits::chalk::_$LT$impl$u20$chalk_solve..RustIrDatabase$LT$chalk_ir..family..ChalkIr$GT$$u20$for$u20$ra_hir_ty..traits..ChalkContext$LT$DB$GT$$GT$::impls_for_trait::h0480fd50100753a9
364.00 ms    1.6%	3.00 ms	 	                                  _$LT$core..iter..adapters..Map$LT$I$C$F$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$::fold::hd960cf6ba4021901
166.00 ms    0.7%	8.00 ms	 	                                  _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$core..ops..drop..Drop$GT$::drop::hb19d10f54021dd07
150.00 ms    0.6%	9.00 ms	 	                                  _$LT$hashbrown..raw..RawTable$LT$T$GT$$u20$as$u20$core..clone..Clone$GT$::clone::h20327c68d9fa8c21
84.00 ms    0.3%	1.00 ms	 	                                  _$LT$chalk_rust_ir..ImplDatum$LT$TF$GT$$u20$as$u20$chalk_solve..clauses..program_clauses..ToProgramClauses$LT$TF$GT$$GT$::to_program_clauses::he8d581f4d271a39f
79.00 ms    0.3%	79.00 ms	 	                                  chalk_rust_ir::Polarity::is_positive::hd3322250332f968e
78.00 ms    0.3%	0 s	 	                                  _$LT$chalk_rust_ir..AssociatedTyDatum$LT$TF$GT$$u20$as$u20$chalk_solve..clauses..program_clauses..ToProgramClauses$LT$TF$GT$$GT$::to_program_clauses::hdbf19ac4bb06dc5f
26.00 ms    0.1%	2.00 ms	 	                                  _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$alloc..vec..SpecExtend$LT$T$C$I$GT$$GT$::from_iter::h43ba57074c7dad4c
26.00 ms    0.1%	1.00 ms	 	                                  _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$alloc..vec..SpecExtend$LT$T$C$I$GT$$GT$::spec_extend::hf5bb697b2fa37bba
22.00 ms    0.0%	0 s	 	                                  <Unknown Address>
17.00 ms    0.0%	3.00 ms	 	                                  free
17.00 ms    0.0%	3.00 ms	 	                                  ra_hir_ty::traits::chalk::_$LT$impl$u20$chalk_solve..RustIrDatabase$LT$chalk_ir..family..ChalkIr$GT$$u20$for$u20$ra_hir_ty..traits..ChalkContext$LT$DB$GT$$GT$::associated_ty_data::h81ff9321b764839d
9.00 ms    0.0%	9.00 ms	 	                                  szone_free_definite_size
9.00 ms    0.0%	6.00 ms	 	                                  free_small
7.00 ms    0.0%	7.00 ms	 	                                  __rdl_dealloc
4.00 ms    0.0%	4.00 ms	 	                                  _$LT$chalk_macros..DEBUG_ENABLED$u20$as$u20$core..ops..deref..Deref$GT$::deref::hfacdb13209b5fe1b
4.00 ms    0.0%	3.00 ms	 	                                  free_tiny
2.00 ms    0.0%	2.00 ms	 	                                  DYLD-STUB$$memcpy
2.00 ms    0.0%	0 s	 	                                  chalk_solve::clauses::match_type_kind::h2a40c751247e71e3
2.00 ms    0.0%	2.00 ms	 	                                  small_free_list_add_ptr
1.00 ms    0.0%	1.00 ms	 	                                  small_free_list_find_by_ptr
1.00 ms    0.0%	1.00 ms	 	                                  _$LT$alloc..vec..IntoIter$LT$T$GT$$u20$as$u20$core..ops..drop..Drop$GT$::drop::h782f389bc30d28b6
1.00 ms    0.0%	1.00 ms	 	                                  core::ptr::real_drop_in_place::hcae11d3a55dab5c5
1.00 ms    0.0%	0 s	 	                                  chalk_solve::split::Split::split_projection::hac00305b91ddf0f6 (.llvm.9692205314475191231)
1.00 ms    0.0%	1.00 ms	 	                                  ra_hir_ty::traits::chalk::_$LT$impl$u20$chalk_solve..RustIrDatabase$LT$chalk_ir..family..ChalkIr$GT$$u20$for$u20$ra_hir_ty..traits..ChalkContext$LT$DB$GT$$GT$::custom_clauses::h9ff37310abd02e4b
1.00 ms    0.0%	1.00 ms	 	                                  DYLD-STUB$$free
4.00 ms    0.0%	0 s	 	                                 <Unknown Address>
1.00 ms    0.0%	1.00 ms	 	                                 szone_free_definite_size
1.54 s    6.8%	8.00 ms	 	                                chalk_engine::logic::_$LT$impl$u20$chalk_engine..forest..Forest$LT$C$GT$$GT$::canonicalize_strand::h270d9d140a2276f8
1.20 s    5.3%	13.00 ms	 	                                chalk_solve::solve::slg::resolvent::_$LT$impl$u20$chalk_engine..context..ResolventOps$LT$chalk_solve..solve..slg..SlgContext$LT$TF$GT$$GT$$u20$for$u20$chalk_solve..solve..slg..TruncatingInferenceTable$LT$TF$GT$$GT$::resolvent_clause::hbcd1bae2e4b07ebb
132.00 ms    0.5%	1.00 ms	 	                                core::ptr::real_drop_in_place::haad0b58dcb0d54ec
31.00 ms    0.1%	0 s	 	                                _$LT$alloc..vec..IntoIter$LT$T$GT$$u20$as$u20$core..ops..drop..Drop$GT$::drop::h782f389bc30d28b6
26.00 ms    0.1%	2.00 ms	 	                                _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$core..clone..Clone$GT$::clone::ha77f0dae400e0c0b
20.00 ms    0.0%	20.00 ms	 	                                _platform_memmove$VARIANT$Haswell
16.00 ms    0.0%	1.00 ms	 	                                _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$core..clone..Clone$GT$::clone::h56eb3c7add71ed97
14.00 ms    0.0%	1.00 ms	 	                                _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$core..ops..drop..Drop$GT$::drop::hb19d10f54021dd07
12.00 ms    0.0%	4.00 ms	 	                                alloc::collections::vec_deque::VecDeque$LT$T$GT$::push_back::h9363a67f3f65e943
11.00 ms    0.0%	3.00 ms	 	                                free_tiny
9.00 ms    0.0%	3.00 ms	 	                                free
7.00 ms    0.0%	0 s	 	                                <Unknown Address>
5.00 ms    0.0%	5.00 ms	 	                                _$LT$chalk_solve..solve..slg..TruncatingInferenceTable$LT$TF$GT$$u20$as$u20$chalk_engine..context..InferenceTable$LT$chalk_solve..solve..slg..SlgContext$LT$TF$GT$$GT$$GT$::into_hh_goal::h93dc13d83f5c3414
2.00 ms    0.0%	2.00 ms	 	                                _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$core..clone..Clone$GT$::clone::hec523ff4582e226b
2.00 ms    0.0%	1.00 ms	 	                                core::ptr::real_drop_in_place::h5a9525201bd9d8c7
1.00 ms    0.0%	1.00 ms	 	                                DYLD-STUB$$memcpy
1.00 ms    0.0%	1.00 ms	 	                                core::ptr::real_drop_in_place::h560f44da1af55f37 (.llvm.7017935223121680485)
1.00 ms    0.0%	1.00 ms	 	                                szone_free_definite_size
1.00 ms    0.0%	1.00 ms	 	                                default_zone_free_definite_size
1.00 ms    0.0%	1.00 ms	 	                                __rust_alloc
93.00 ms    0.4%	0 s	 	                               chalk_engine::tables::Tables$LT$C$GT$::insert::h7736d6c431b2e7e4
31.00 ms    0.1%	0 s	 	                               chalk_solve::infer::InferenceTable$LT$TF$GT$::from_canonical::hdb112e6162412ccd
30.00 ms    0.1%	14.00 ms	 	                               chalk_engine::tables::Tables$LT$C$GT$::index_of::he5913f10d9ae4771
28.00 ms    0.1%	0 s	 	                               core::ptr::real_drop_in_place::h12745c8101439c70
13.00 ms    0.0%	0 s	 	                               _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$core..clone..Clone$GT$::clone::hd123cbaefed998ba
11.00 ms    0.0%	3.00 ms	 	                               _$LT$chalk_ir..Goal$LT$TF$GT$$u20$as$u20$core..clone..Clone$GT$::clone::h6b0092bed962c98f (.llvm.15219803596393920156)
4.00 ms    0.0%	0 s	 	                               <Unknown Address>
2.00 ms    0.0%	1.00 ms	 	                               free
2.00 ms    0.0%	2.00 ms	 	                               _platform_memmove$VARIANT$Haswell
1.00 ms    0.0%	1.00 ms	 	                               core::ptr::real_drop_in_place::haad0b58dcb0d54ec
1.00 ms    0.0%	1.00 ms	 	                               _$LT$chalk_macros..DEBUG_ENABLED$u20$as$u20$core..ops..deref..Deref$GT$::deref::hfacdb13209b5fe1b
311.00 ms    1.3%	4.00 ms	 	                              _$LT$chalk_solve..solve..slg..SlgContextOps$LT$TF$GT$$u20$as$u20$chalk_engine..context..ContextOps$LT$chalk_solve..solve..slg..SlgContext$LT$TF$GT$$GT$$GT$::instantiate_ex_clause::h8b7159a3c0abe826
224.00 ms    0.9%	4.00 ms	 	                              chalk_engine::logic::_$LT$impl$u20$chalk_engine..forest..Forest$LT$C$GT$$GT$::unwind_stack::h2f8b83af1cf8637c
158.00 ms    0.7%	1.00 ms	 	                              chalk_solve::infer::ucanonicalize::_$LT$impl$u20$chalk_solve..infer..InferenceTable$LT$TF$GT$$GT$::u_canonicalize::h6b3e5abf876212e7
114.00 ms    0.5%	2.00 ms	 	                              core::ptr::real_drop_in_place::hb1337e974fede814
90.00 ms    0.3%	1.00 ms	 	                              chalk_engine::logic::_$LT$impl$u20$chalk_engine..forest..Forest$LT$C$GT$$GT$::canonicalize_strand::h270d9d140a2276f8
72.00 ms    0.3%	4.00 ms	 	                              chalk_solve::infer::canonicalize::_$LT$impl$u20$chalk_solve..infer..InferenceTable$LT$TF$GT$$GT$::canonicalize::h69669291605111ce
50.00 ms    0.2%	0 s	 	                              chalk_engine::logic::_$LT$impl$u20$chalk_engine..forest..Forest$LT$C$GT$$GT$::merge_answer_into_strand::h77211240cd2d698d
49.00 ms    0.2%	0 s	 	                              chalk_engine::table::Table$LT$C$GT$::mark_floundered::h4623a23bec20cf27
38.00 ms    0.1%	3.00 ms	 	                              chalk_solve::solve::truncate::truncate::h7fa5497587289943
18.00 ms    0.0%	0 s	 	                              core::ptr::real_drop_in_place::h12745c8101439c70 (.llvm.12484288865420280758)
16.00 ms    0.0%	16.00 ms	 	                              _platform_memmove$VARIANT$Haswell
13.00 ms    0.0%	2.00 ms	 	                              free_tiny
9.00 ms    0.0%	2.00 ms	 	                              free
6.00 ms    0.0%	0 s	 	                              chalk_engine::logic::_$LT$impl$u20$chalk_engine..forest..Forest$LT$C$GT$$GT$::clear_strands_after_cycle::ha80f78fc522cf7ef
6.00 ms    0.0%	0 s	 	                              _$LT$alloc..collections..vec_deque..VecDeque$LT$T$GT$$u20$as$u20$core..ops..drop..Drop$GT$::drop::h56b6b59a278434ab
6.00 ms    0.0%	6.00 ms	 	                              core::ptr::real_drop_in_place::h560f44da1af55f37 (.llvm.12484288865420280758)
6.00 ms    0.0%	0 s	 	                              <Unknown Address>
5.00 ms    0.0%	0 s	 	                              chalk_engine::logic::_$LT$impl$u20$chalk_engine..forest..Forest$LT$C$GT$$GT$::flounder_subgoal::hb79be2740255cb5a
4.00 ms    0.0%	2.00 ms	 	                              core::ptr::real_drop_in_place::h02533b82efae4857
3.00 ms    0.0%	0 s	 	                              _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$core..clone..Clone$GT$::clone::h33cc6e9538d568ee
3.00 ms    0.0%	3.00 ms	 	                              chalk_engine::stack::Stack$LT$C$GT$::is_active::h9c59e8aee1afd8b9
1.00 ms    0.0%	1.00 ms	 	                              _$LT$chalk_macros..DEBUG_ENABLED$u20$as$u20$core..ops..deref..Deref$GT$::deref::hfacdb13209b5fe1b
1.00 ms    0.0%	1.00 ms	 	                              DYLD-STUB$$memcpy
1.00 ms    0.0%	0 s	 	                              chalk_engine::stack::Stack$LT$C$GT$::pop_and_adjust_depth::h1a5860bea8cfb591 (.llvm.15448677400207065130)
1.00 ms    0.0%	1.00 ms	 	                              chalk_macros::Indent::new::h1ed049dec3cf72e8
1.00 ms    0.0%	0 s	 	                              core::ptr::real_drop_in_place::hc802de20083df5ce
6.00 ms    0.0%	0 s	 	                             <Unknown Address>
54.00 ms    0.2%	0 s	 	                           chalk_engine::logic::_$LT$impl$u20$chalk_engine..forest..Forest$LT$C$GT$$GT$::get_or_create_table_for_ucanonical_goal::ha2452aa1765187bb (.llvm.15219803596393920156)
53.00 ms    0.2%	0 s	 	                            chalk_engine::logic::_$LT$impl$u20$chalk_engine..forest..Forest$LT$C$GT$$GT$::push_initial_strands_instantiated::hbcb2aa5857d1b1c2
28.00 ms    0.1%	0 s	 	                             _$LT$chalk_solve..solve..slg..SlgContextOps$LT$TF$GT$$u20$as$u20$chalk_engine..context..ContextOps$LT$chalk_solve..solve..slg..SlgContext$LT$TF$GT$$GT$$GT$::program_clauses::h7d3946d01e66280e
28.00 ms    0.1%	0 s	 	                              chalk_solve::clauses::program_clauses_for_goal::h100574e82ab49014
9.00 ms    0.0%	1.00 ms	 	                               _$LT$chalk_rust_ir..AssociatedTyValue$LT$TF$GT$$u20$as$u20$chalk_solve..clauses..program_clauses..ToProgramClauses$LT$TF$GT$$GT$::to_program_clauses::h1eb26983f36338ac
8.00 ms    0.0%	0 s	 	                                chalk_solve::clauses::builder::ClauseBuilder$LT$TF$GT$::push_binders::h972d2cd1888a132a
3.00 ms    0.0%	0 s	 	                                 malloc
1.00 ms    0.0%	0 s	 	                                 chalk_solve::split::Split::impl_parameters_and_projection_from_associated_ty_value::hcd0a1bea38ea81d4
1.00 ms    0.0%	0 s	 	                                 chalk_ir::Binders$LT$T$GT$::substitute::h77e1c6b1b7bf8956
1.00 ms    0.0%	1.00 ms	 	                                 _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$alloc..vec..SpecExtend$LT$T$C$I$GT$$GT$::from_iter::he67ed9f4974c6a2e
1.00 ms    0.0%	1.00 ms	 	                                 _platform_memmove$VARIANT$Haswell
1.00 ms    0.0%	0 s	 	                                 alloc::vec::Vec$LT$T$GT$::extend_from_slice::h9f9ec3887999902f
1.00 ms    0.0%	0 s	 	                                  _$LT$core..iter..adapters..Cloned$LT$I$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$::fold::h55b571b727a4bd52
1.00 ms    0.0%	0 s	 	                                   _$LT$chalk_ir..Ty$LT$TF$GT$$u20$as$u20$core..clone..Clone$GT$::clone::h750f767d155cbf68 (.llvm.13130049199905610347)
1.00 ms    0.0%	0 s	 	                                    malloc
1.00 ms    0.0%	0 s	 	                                     malloc_zone_malloc
1.00 ms    0.0%	0 s	 	                                      szone_malloc_should_clear
1.00 ms    0.0%	1.00 ms	 	                                       tiny_malloc_should_clear
8.00 ms    0.0%	0 s	 	                               _$LT$chalk_rust_ir..ImplDatum$LT$TF$GT$$u20$as$u20$chalk_solve..clauses..program_clauses..ToProgramClauses$LT$TF$GT$$GT$::to_program_clauses::he8d581f4d271a39f
8.00 ms    0.0%	0 s	 	                                chalk_solve::clauses::builder::ClauseBuilder$LT$TF$GT$::push_binders::ha747159d08cc039c
5.00 ms    0.0%	0 s	 	                                 chalk_ir::Binders$LT$T$GT$::substitute::h14488951bab46f2c
5.00 ms    0.0%	0 s	 	                                  _$LT$$LP$A$C$$u20$B$RP$$u20$as$u20$chalk_ir..fold..Fold$LT$TF$C$TTF$GT$$GT$::fold_with::h50981795c882fd68
2.00 ms    0.0%	0 s	 	                                 _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$alloc..vec..SpecExtend$LT$T$C$I$GT$$GT$::from_iter::h8627dc696f1ec884
1.00 ms    0.0%	0 s	 	                                 alloc::raw_vec::RawVec$LT$T$C$A$GT$::reserve::h474aa6550149bc73
5.00 ms    0.0%	0 s	 	                               ra_hir_ty::traits::chalk::_$LT$impl$u20$chalk_solve..RustIrDatabase$LT$chalk_ir..family..ChalkIr$GT$$u20$for$u20$ra_hir_ty..traits..ChalkContext$LT$DB$GT$$GT$::impl_datum::h0d797a9d82263194
3.00 ms    0.0%	0 s	 	                               alloc::vec::Vec$LT$T$GT$::retain::h7528e9b48b4e7922
1.00 ms    0.0%	1.00 ms	 	                               _$LT$hashbrown..raw..RawTable$LT$T$GT$$u20$as$u20$core..clone..Clone$GT$::clone::h20327c68d9fa8c21
1.00 ms    0.0%	1.00 ms	 	                               __rust_alloc
1.00 ms    0.0%	0 s	 	                               ra_hir_ty::traits::chalk::_$LT$impl$u20$chalk_solve..RustIrDatabase$LT$chalk_ir..family..ChalkIr$GT$$u20$for$u20$ra_hir_ty..traits..ChalkContext$LT$DB$GT$$GT$::associated_ty_value::h0b10268faf2a3355
14.00 ms    0.0%	0 s	 	                             chalk_engine::logic::_$LT$impl$u20$chalk_engine..forest..Forest$LT$C$GT$$GT$::canonicalize_strand::h270d9d140a2276f8
8.00 ms    0.0%	1.00 ms	 	                             chalk_solve::solve::slg::resolvent::_$LT$impl$u20$chalk_engine..context..ResolventOps$LT$chalk_solve..solve..slg..SlgContext$LT$TF$GT$$GT$$u20$for$u20$chalk_solve..solve..slg..TruncatingInferenceTable$LT$TF$GT$$GT$::resolvent_clause::hbcd1bae2e4b07ebb
1.00 ms    0.0%	0 s	 	                             _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$core..ops..drop..Drop$GT$::drop::hb19d10f54021dd07
1.00 ms    0.0%	0 s	 	                             alloc::collections::vec_deque::VecDeque$LT$T$GT$::push_back::h9363a67f3f65e943
1.00 ms    0.0%	0 s	 	                             free
1.00 ms    0.0%	0 s	 	                            chalk_solve::infer::InferenceTable$LT$TF$GT$::from_canonical::hdb112e6162412ccd
2.00 ms    0.0%	1.00 ms	 	                     salsa::derived::slot::Slot$LT$DB$C$Q$C$MP$GT$::register_with_in_progress_thread::h89e3e702267b57de
2.00 ms    0.0%	0 s	 	                 rowan::cursor::SyntaxNode::next_sibling::he6cc276793710241
1.00 ms    0.0%	0 s	 	                 core::fmt::write::h4f487e714088986d
1.00 ms    0.0%	1.00 ms	 	                 _$LT$core..iter..adapters..Map$LT$I$C$F$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$::try_fold::h644acbe8b3eb520d
1.00 ms    0.0%	1.00 ms	 	                 _$LT$alloc..rc..Rc$LT$T$GT$$u20$as$u20$core..ops..drop..Drop$GT$::drop::hbf549a176472e4f5
311.00 ms    1.3%	0 s	 	           ra_lsp_server::main_loop::handlers::publish_decorations::h303299ee3d93ff3e
16.00 ms    0.0%	0 s	 	           serde_json::value::to_value::h7bb82279f3ceaf15
1.00 ms    0.0%	0 s	 	           serde_json::value::to_value::ha84f5901cd7f78e8
1.71 s    7.5%	0 s	 	 std::sys::unix::thread::Thread::new::thread_start::h785fcf8d81bd0f52  0x4a6dd
462.00 ms    2.0%	0 s	 	 std::sys::unix::thread::Thread::new::thread_start::h785fcf8d81bd0f52  0x4a6e0
398.00 ms    1.7%	0 s	 	 std::sys::unix::thread::Thread::new::thread_start::h785fcf8d81bd0f52  0x4a676
386.00 ms    1.7%	0 s	 	 std::sys::unix::thread::Thread::new::thread_start::h785fcf8d81bd0f52  0x4a6db
386.00 ms    1.7%	0 s	 	 std::sys::unix::thread::Thread::new::thread_start::h785fcf8d81bd0f52  0x4a6de
383.00 ms    1.6%	0 s	 	 std::sys::unix::thread::Thread::new::thread_start::h785fcf8d81bd0f52  0x4a6dc
356.00 ms    1.5%	0 s	 	 std::sys::unix::thread::Thread::new::thread_start::h785fcf8d81bd0f52  0x4a6df
348.00 ms    1.5%	0 s	 	 std::sys::unix::thread::Thread::new::thread_start::h785fcf8d81bd0f52  0x4a6e1
8.00 ms    0.0%	0 s	 	 Main Thread  0x4a615
7.00 ms    0.0%	0 s	 	 std::sys::unix::thread::Thread::new::thread_start::h785fcf8d81bd0f52  0x4a677
```

</details>

**Comment by Matthias247** [comment]
It seems like it actually doesn't crash, and the CPU usage recovers after some time. You can see this in the following profiler screenshot: At first there is an analysis running on one of the worker threads, which finishes after some time. CPU usage recovers. Then I click a little bit around in VsCode, which leads to the smaller spikes on all threads between 17s and 30s. Then I did something (goto definition?), which caused again a 100% workload to start on another worker thread, which starts at 30s and only finishes at 1:25.

<img width="1280" alt="Screen Shot 2019-12-09 at 15 41 05" src="https://user-images.githubusercontent.com/6330334/70482448-56b1cc80-1a9a-11ea-8ebe-5378546013a5.png">

**Comment by Dowwie** [comment]
Unfortunately, rust-analyzer consumed 100% cpu and memory again for me (debian stretch, latest VSC).  It happens when I load a rust file from the command line into VSC and the file isn't part of the current workspace loaded in VSC at the time.

**Comment by jan-auer** [comment]
I might have a project for the maintainers to reproduce this: https://github.com/getsentry/relay

rust-analyzer has worked fine before upgrading to `2020-05-04` (VSCode extension version `0.2.159`). After this, RA has started to consume constant 100% CPU (1 core) and leaks about 50MB/s, indefinitely. I've not seen it recover even after 40 minutes. I'm also noticing that it's not running the formatter on save and just gets stuck.

I'm running with default settings, but the problem even persists when disabling "all targets" or "all features".

**Comment by matklad** [comment]
@jan-auer opened https://github.com/rust-analyzer/rust-analyzer/issues/4463 for this specific issue. 

I think this one we should close -- it seems like we at various times had a bunch of different issues with similar symptoms, and having them all in one issue is just confusing.
